### ISSUE TITLE
Python 3.10.0beta test_writers.py failures: missing quoting

### ISSUE THREAD

**Issue by mgorny** [xflr6/csv23#6]
The tests are failing with Python 3.10.0b1. Tested with git master (1677f82a8d016e95e9cb8a6933c7fc8222ec6018 and 7f10b99666e1c8f13906498e9e020d6682335299), get the same results, so I'm wondering if the py3.10 fix works.

```
FAILED tests/test_stdlib_csv.py::test_csv_roundtrip[spam\\eggs-0-"-\\-True] - AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']
FAILED tests/test_stdlib_csv.py::test_csv_roundtrip[spam\\eggs-0-"-\\-False] - AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']
FAILED tests/test_writers.py::test_open_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\e...
FAILED tests/test_writers.py::test_open_writer[utf-16-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\...
FAILED tests/test_writers.py::test_open_writer[utf-8-sig-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\...
FAILED tests/test_writers.py::test_open_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\...
FAILED tests/test_writers.py::test_writer[False-row5-fmtparams5-"spam\\\\eggs"\r\n] - AssertionError: assert False
FAILED tests/test_writers.py::test_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg......
FAILED tests/test_writers.py::test_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg.....
```

Full output:
```pytb
$ tox -e py310
GLOB sdist-make: /tmp/csv23/setup.py
py310 inst-nodeps: /tmp/csv23/.tox/.tmp/package/1/csv23-0.3.3.dev0.zip
py310 installed: attrs==21.2.0,coverage==5.5,csv23 @ file:///tmp/csv23/.tox/.tmp/package/1/csv23-0.3.3.dev0.zip,iniconfig==1.1.1,mock==4.0.3,packaging==20.9,pluggy==0.13.1,py==1.10.0,pyparsing==2.4.7,pytest==6.2.4,pytest-cov==2.12.0,pytest-mock==3.6.1,toml==0.10.2
py310 run-test-pre: PYTHONHASHSEED='3528200982'
py310 run-test: commands[0] | ./run-tests.py
============================================================== test session starts ===============================================================
platform linux -- Python 3.10.0b1, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
cachedir: .tox/py310/.pytest_cache
rootdir: /tmp/csv23, configfile: setup.cfg, testpaths: csv23, tests
plugins: cov-2.12.0, mock-3.6.1
collected 339 items                                                                                                                              

csv23/__init__.py s.                                                                                                                       [  0%]
csv23/_common.py .                                                                                                                         [  0%]
csv23/extras.py .                                                                                                                          [  1%]
csv23/openers.py ss                                                                                                                        [  1%]
csv23/readers.py .                                                                                                                         [  2%]
csv23/shortcuts.py ..                                                                                                                      [  2%]
csv23/writers.py .                                                                                                                         [  2%]
tests/test_common.py ........ssss.                                                                                                         [  6%]
tests/test_dialect.py ..............                                                                                                       [ 10%]
tests/test_encoding.py ......                                                                                                              [ 12%]
tests/test_extras.py ............                                                                                                          [ 16%]
tests/test_init.py .........                                                                                                               [ 18%]
tests/test_openers.py ...                                                                                                                  [ 19%]
tests/test_readers.py .......x.......x.......x.....s.x.......x.......x.....s.x                                                             [ 36%]
tests/test_rowtype.py ..........                                                                                                           [ 39%]
tests/test_shortcuts.py ss........................................                                                                         [ 51%]
tests/test_stdlib_csv.py ..................................................................................FF............                  [ 79%]
tests/test_stdlib_truncate.py s.                                                                                                           [ 80%]
tests/test_workarounds.py ..                                                                                                               [ 81%]
tests/test_writers.py .....F.......F.......F......sF...............F.......F......sF..                                                     [100%]

==================================================================== FAILURES ====================================================================
___________________________________________________ test_csv_roundtrip[spam\\eggs-0-"-\\-True] ___________________________________________________

py2 = False, quoting = 0, quotechar = '"', escapechar = '\\', doublequote = True, value = 'spam\\eggs'

    @pytest.mark.parametrize(', '.join(FORMATS[0]), FORMATS[1:])
    @pytest.mark.parametrize('value', VALUES)
    def test_csv_roundtrip(py2, quoting, quotechar, escapechar, doublequote, value):
        csv_kwargs = {'quoting': quoting, 'quotechar': quotechar,
                      'escapechar': escapechar, 'doublequote': doublequote}
        row = expected = [value]
    
        with io.BytesIO() if py2 else io.StringIO(newline='') as f:
            if quoting != csv.QUOTE_NONE and quotechar is None:
                with pytest.raises(TypeError, match=r'quotechar must be set'):
                    csv.writer(f, **csv_kwargs)
                return
    
            writer = csv.writer(f, **csv_kwargs)
    
            if escapechar is None and (quoting == csv.QUOTE_NONE or not doublequote):
                need_escape = [',', '\r', '\n'] if quoting == csv.QUOTE_NONE else []
                if quotechar is not None:
                    need_escape.append(quotechar)
                if any(n in value for n in need_escape):
                    with pytest.raises(Exception, match=r'need to escape'):
                        writer.writerow(row)
                    return
    
            writer.writerow(row)
            f.seek(0)
            row = next(csv.reader(f, **csv_kwargs))
    
        if quoting != csv.QUOTE_NONE and escapechar is not None and escapechar in value:
            # writer fails to escape escapechar
>           assert row != expected
E           AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']

tests/test_stdlib_csv.py:67: AssertionError
__________________________________________________ test_csv_roundtrip[spam\\eggs-0-"-\\-False] ___________________________________________________

py2 = False, quoting = 0, quotechar = '"', escapechar = '\\', doublequote = False, value = 'spam\\eggs'

    @pytest.mark.parametrize(', '.join(FORMATS[0]), FORMATS[1:])
    @pytest.mark.parametrize('value', VALUES)
    def test_csv_roundtrip(py2, quoting, quotechar, escapechar, doublequote, value):
        csv_kwargs = {'quoting': quoting, 'quotechar': quotechar,
                      'escapechar': escapechar, 'doublequote': doublequote}
        row = expected = [value]
    
        with io.BytesIO() if py2 else io.StringIO(newline='') as f:
            if quoting != csv.QUOTE_NONE and quotechar is None:
                with pytest.raises(TypeError, match=r'quotechar must be set'):
                    csv.writer(f, **csv_kwargs)
                return
    
            writer = csv.writer(f, **csv_kwargs)
    
            if escapechar is None and (quoting == csv.QUOTE_NONE or not doublequote):
                need_escape = [',', '\r', '\n'] if quoting == csv.QUOTE_NONE else []
                if quotechar is not None:
                    need_escape.append(quotechar)
                if any(n in value for n in need_escape):
                    with pytest.raises(Exception, match=r'need to escape'):
                        writer.writerow(row)
                    return
    
            writer.writerow(row)
            f.seek(0)
            row = next(csv.reader(f, **csv_kwargs))
    
        if quoting != csv.QUOTE_NONE and escapechar is not None and escapechar in value:
            # writer fails to escape escapechar
>           assert row != expected
E           AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']

tests/test_stdlib_csv.py:67: AssertionError
___________________________________________ test_open_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] ___________________________________________

py2 = False, filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_open_writer_utf_8_row5_fm0/spam.csv'), encoding = 'utf-8'
row = ['spam\\eggs'], fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_open_writer(py2, filepath, encoding, row, fmtparams, expected, n=12):
        try:
            expected.encode(encoding)
        except UnicodeEncodeError:
            pytest.skip('impossible combination of row and encoding')
    
        write_n = len(expected.encode(encoding) if py2 and is_8bit_clean(encoding) else expected)
    
        filename = str(filepath)
    
        with open_writer(filename, encoding=encoding, **fmtparams) as w:
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:55: AssertionError
__________________________________________ test_open_writer[utf-16-row5-fmtparams5-"spam\\\\eggs"\r\n] ___________________________________________

py2 = False, filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_open_writer_utf_16_row5_f0/spam.csv'), encoding = 'utf-16'
row = ['spam\\eggs'], fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_open_writer(py2, filepath, encoding, row, fmtparams, expected, n=12):
        try:
            expected.encode(encoding)
        except UnicodeEncodeError:
            pytest.skip('impossible combination of row and encoding')
    
        write_n = len(expected.encode(encoding) if py2 and is_8bit_clean(encoding) else expected)
    
        filename = str(filepath)
    
        with open_writer(filename, encoding=encoding, **fmtparams) as w:
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:55: AssertionError
_________________________________________ test_open_writer[utf-8-sig-row5-fmtparams5-"spam\\\\eggs"\r\n] _________________________________________

py2 = False, filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_open_writer_utf_8_sig_row5/spam.csv'), encoding = 'utf-8-sig'
row = ['spam\\eggs'], fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_open_writer(py2, filepath, encoding, row, fmtparams, expected, n=12):
        try:
            expected.encode(encoding)
        except UnicodeEncodeError:
            pytest.skip('impossible combination of row and encoding')
    
        write_n = len(expected.encode(encoding) if py2 and is_8bit_clean(encoding) else expected)
    
        filename = str(filepath)
    
        with open_writer(filename, encoding=encoding, **fmtparams) as w:
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:55: AssertionError
__________________________________________ test_open_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] ___________________________________________

py2 = False, filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_open_writer_latin9_row5_f0/spam.csv'), encoding = 'latin9'
row = ['spam\\eggs'], fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_open_writer(py2, filepath, encoding, row, fmtparams, expected, n=12):
        try:
            expected.encode(encoding)
        except UnicodeEncodeError:
            pytest.skip('impossible combination of row and encoding')
    
        write_n = len(expected.encode(encoding) if py2 and is_8bit_clean(encoding) else expected)
    
        filename = str(filepath)
    
        with open_writer(filename, encoding=encoding, **fmtparams) as w:
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:55: AssertionError
_____________________________________________ test_writer[False-row5-fmtparams5-"spam\\\\eggs"\r\n] ______________________________________________

mocker = <pytest_mock.plugin.MockerFixture object at 0x7ff8b28362c0>, py2 = False
filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_writer_False_row5_fmtpara0/spam.csv'), inner_encoding = False, row = ['spam\\eggs']
fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_writer(mocker, py2, filepath, inner_encoding,
                    row, fmtparams, expected, n=12):
        encoding = inner_encoding
        if encoding is False:
            if py2:
                expected_type = UnicodeTextWriter
            elif (fmtparams.get('escapechar')
                  and fmtparams.get('quoting', csv.QUOTE_MINIMAL) != csv.QUOTE_NONE):
                expected_type = _UnicodeTextWriter
            else:
                expected_type = type(csv.writer(mocker.mock_open()()))
            write_n = len(expected)
            file_encoding = 'utf-8'
            open_kwargs = {'mode': 'w', 'encoding': file_encoding, 'newline': ''}
        else:
            expected_type = UnicodeBytesWriter
            try:
                write_n = len(expected.encode(encoding))
            except UnicodeEncodeError:
                pytest.skip('impossible combination of row and encoding')
            file_encoding = encoding
            open_kwargs = {'mode': 'wb'}
    
        filename = str(filepath)
    
        with io.open(filename, **open_kwargs) as f:
            w = writer(f, encoding=encoding, **fmtparams)
>           assert isinstance(w, expected_type)
E           AssertionError: assert False
E            +  where False = isinstance(<_csv.writer object at 0x7ff8b2697b20>, <class 'csv23.writers._UnicodeTextWriter'>)

tests/test_writers.py:99: AssertionError
_____________________________________________ test_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] ______________________________________________

mocker = <pytest_mock.plugin.MockerFixture object at 0x7ff8b23e1a20>, py2 = False
filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_writer_utf_8_row5_fmtpara0/spam.csv'), inner_encoding = 'utf-8', row = ['spam\\eggs']
fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_writer(mocker, py2, filepath, inner_encoding,
                    row, fmtparams, expected, n=12):
        encoding = inner_encoding
        if encoding is False:
            if py2:
                expected_type = UnicodeTextWriter
            elif (fmtparams.get('escapechar')
                  and fmtparams.get('quoting', csv.QUOTE_MINIMAL) != csv.QUOTE_NONE):
                expected_type = _UnicodeTextWriter
            else:
                expected_type = type(csv.writer(mocker.mock_open()()))
            write_n = len(expected)
            file_encoding = 'utf-8'
            open_kwargs = {'mode': 'w', 'encoding': file_encoding, 'newline': ''}
        else:
            expected_type = UnicodeBytesWriter
            try:
                write_n = len(expected.encode(encoding))
            except UnicodeEncodeError:
                pytest.skip('impossible combination of row and encoding')
            file_encoding = encoding
            open_kwargs = {'mode': 'wb'}
    
        filename = str(filepath)
    
        with io.open(filename, **open_kwargs) as f:
            w = writer(f, encoding=encoding, **fmtparams)
            assert isinstance(w, expected_type)
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=file_encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:106: AssertionError
_____________________________________________ test_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] _____________________________________________

mocker = <pytest_mock.plugin.MockerFixture object at 0x7ff8b23e22f0>, py2 = False
filepath = PosixPath('/tmp/pytest-of-mgorny/pytest-8/test_writer_latin9_row5_fmtpar0/spam.csv'), inner_encoding = 'latin9', row = ['spam\\eggs']
fmtparams = {'escapechar': '\\', 'quoting': 0}, expected = '"spam\\\\eggs"\r\n', n = 12

    @pytest.mark.parametrize('row, fmtparams, expected', ROW_FORMAT_LINE)
    def test_writer(mocker, py2, filepath, inner_encoding,
                    row, fmtparams, expected, n=12):
        encoding = inner_encoding
        if encoding is False:
            if py2:
                expected_type = UnicodeTextWriter
            elif (fmtparams.get('escapechar')
                  and fmtparams.get('quoting', csv.QUOTE_MINIMAL) != csv.QUOTE_NONE):
                expected_type = _UnicodeTextWriter
            else:
                expected_type = type(csv.writer(mocker.mock_open()()))
            write_n = len(expected)
            file_encoding = 'utf-8'
            open_kwargs = {'mode': 'w', 'encoding': file_encoding, 'newline': ''}
        else:
            expected_type = UnicodeBytesWriter
            try:
                write_n = len(expected.encode(encoding))
            except UnicodeEncodeError:
                pytest.skip('impossible combination of row and encoding')
            file_encoding = encoding
            open_kwargs = {'mode': 'wb'}
    
        filename = str(filepath)
    
        with io.open(filename, **open_kwargs) as f:
            w = writer(f, encoding=encoding, **fmtparams)
            assert isinstance(w, expected_type)
            written = w.writerow(row)
            w.writerows([row] * (n - 1))
    
        with io.open(filename, encoding=file_encoding, newline='') as f:
            line = f.read()
    
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
...       - "spam\\eggs"
E         
E         ...Full output truncated (30 lines hidden), use '-vv' to show

tests/test_writers.py:106: AssertionError

----------- coverage: platform linux, python 3.10.0-beta-1 -----------
Name                    Stmts   Miss  Cover
-------------------------------------------
csv23/__init__.py          28      0   100%
csv23/_common.py           33      6    82%
csv23/_dispatch.py         29      0   100%
csv23/_workarounds.py      22      1    95%
csv23/dialects.py          28      8    71%
csv23/extras.py            59      2    97%
csv23/openers.py           35      6    83%
csv23/readers.py           59     16    73%
csv23/shortcuts.py        107      8    93%
csv23/writers.py           90     33    63%
-------------------------------------------
TOTAL                     490     80    84%
Coverage HTML written to dir htmlcov

============================================================ short test summary info =============================================================
FAILED tests/test_stdlib_csv.py::test_csv_roundtrip[spam\\eggs-0-"-\\-True] - AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']
FAILED tests/test_stdlib_csv.py::test_csv_roundtrip[spam\\eggs-0-"-\\-False] - AssertionError: assert ['spam\\eggs'] != ['spam\\eggs']
FAILED tests/test_writers.py::test_open_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\e...
FAILED tests/test_writers.py::test_open_writer[utf-16-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\...
FAILED tests/test_writers.py::test_open_writer[utf-8-sig-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\...
FAILED tests/test_writers.py::test_open_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\...
FAILED tests/test_writers.py::test_writer[False-row5-fmtparams5-"spam\\\\eggs"\r\n] - AssertionError: assert False
FAILED tests/test_writers.py::test_writer[utf-8-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg......
FAILED tests/test_writers.py::test_writer[latin9-row5-fmtparams5-"spam\\\\eggs"\r\n] - assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg.....
============================================== 9 failed, 309 passed, 14 skipped, 7 xfailed in 1.47s ==============================================
ERROR: InvocationError for command /tmp/csv23/run-tests.py (exited with code 1)
____________________________________________________________________ summary _____________________________________________________________________
ERROR:   py310: commands failed

```

**Comment by xflr6** [comment]
Thanks for the report.

Some of it seems to be adapting the test assertions, xfails, etc. Need to check if there is also a problem here:

https://github.com/xflr6/csv23/blob/1677f82a8d016e95e9cb8a6933c7fc8222ec6018/csv23/writers.py#L74-L76

**Comment by xflr6** [comment]
There is a chance that this test show behavioural change in Python (not sure if it would be related to the fix of https://bugs.python.org/issue12178):

```python
In [1]: import csv
   ...: 
   ...: csv.QUOTE_MINIMAL
Out[1]: 0
```

```python
>       assert line == expected * n
E       assert 'spam\\\\eggs...m\\\\eggs\r\n' == '"spam\\\\egg...\\\\eggs"\r\n'
E         - "spam\\eggs"
E         ? -          -
E         + spam\\eggs
```
https://docs.python.org/3/library/csv.html#csv.QUOTE_MINIMAL

> Instructs writer objects to only quote those fields which contain special characters such as delimiter,
> quotechar or any of the characters in lineterminator

**Comment by mgorny** [comment]
Since csv23 is now a dependency of pip, this is quite a blocker for py3.10 porting in Gentoo. Could you tell me whether the test failures are ok to ignore for now (i.e. the new output is still valid, just different)?

**Comment by xflr6** [comment]
> Since csv23 is now a dependency of pip, this is quite a blocker for py3.10 porting in Gentoo. 

Oh, that woud be news to me: cannot find this in [pip's setup.py](https://github.com/pypa/pip/blob/57be6a77c57ab5d512371b5c48d508a7620c3217/setup.py#L28), do you maybe have a reference for this?

**Comment by mgorny** [comment]
Ah, they've removed it!

```
commit 36602b7977adff2a7d123bb273307d1cd099257e
Merge: 609d9d4a6 3cfb0ca1e
Author: Pradyun Gedam <3275593+pradyunsg@users.noreply.github.com>
Date:   2020-12-27 18:05:54 +0100

    Merge pull request #9374 from jdufresne/csv23
    
    Drop csv23 and enum34 in favor of stdlib
```

Thanks for checking! This means I've got a stale dependency and can forget about this problem for the time being ;-).

**Comment by xflr6** [comment]
Reproduced adding Python 3.10.0-beta to test matrix.

https://github.com/xflr6/csv23/runs/3059197134?check_suite_focus=true

**Comment by xflr6** [comment]
Okay, looks like this is about adapting the test branches to reflect the upstream fix.

**Comment by xflr6** [comment]
Fixed the test logic in 30b9c95d8b5ef93e48aa444bb2ac9126af5ec5ea.

Remaining 7 failiures are an undocumented behavioural change AFAIU.

Reported in https://bugs.python.org/issue12178#msg397440 as follows:

Python 3.9 quotes values with escapechar:

```python
import csv
import io

kwargs = {'escapechar': '\\'}

value = 'spam\\eggs'

print(value)

with io.StringIO() as buf:
    writer = csv.writer(buf, **kwargs)
    writer.writerow([value])
    line = buf.getvalue()

print(line.strip())

with io.StringIO(line) as buf:
    reader = csv.reader(buf, **kwargs)
    (new_value,), = reader

print(new_value)
spam\eggs
"spam\eggs"
spameggs
```

- quotes string with escapechar
- fails to double the escapechar (this bug)

https://docs.python.org/3/library/csv.html#csv.QUOTE_MINIMAL

> only quote those fields which contain special characters
> such as delimiter, quotechar or any of the characters in lineterminator.

This seems incorrect because escapechar is not mentioned (but at the same time it says 'such as') and maybe better matching the name 'minimal' (or one might expect 'more' quoting as a better default).

Python 3.10:

https://github.com/python/cpython/blob/5c0eed7375fdd791cc5e19ceabfab4170ad44062/Lib/test/test_csv.py#L207-L208

See also https://github.com/xflr6/csv23/actions/runs/1027687524

**Comment by xflr6** [comment]
Adapted the test expectations in 8d56b3ad31fe1c747e203e01680056b8cc926ed1, fixed assertion in 1a882a0e8ca153db30e0b4bc3763477c407438a9, and demoted to note in e96ee9c9d612b8070f93ccd9f570c8a5b6e0964b.

Closing for now, pending upstream answer.

**Comment by xflr6** [comment]
Opened a new upstream issue: https://bugs.python.org/issue44861
